### PR TITLE
Serverassignment

### PR DESCRIPTION
**Pull request by thottel64**
Here is my first attempt at the Server Assignment. I struggled a bit with Delete and Put  requests since we never covered them in class, but I tried my best with some googling.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+	//step 3 unmarshal the slice of bytes into a struct
+	err = json.Unmarshal(byteslice, &product)
+	if err != nil {
+		log.Fatalln(err)
+	}
+	// step 4 print out the data
+	log.Printf("%s", response.Status)
+	fmt.Println(len(product))
+	o := product[1]
+	fmt.Println(o)
```
Comment (jboursiquot) [inline main.go]: No need to print out here. You can use the `log` package if you need to emit something for troubleshooting purposes.

File: main.go [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+		fmt.Fprintf(w, "ParseForm() err: %v", err)
+		return
+	}
+	fmt.Fprintf(w, "Post from website! r.PostFrom = %v\n", r.PostForm)
+	// these fields need to be changed to match our JSON struct
+	ID := r.FormValue("ID")
+	Name := r.FormValue("Name")
+	Description := r.FormValue("Description")
+	Price := r.FormValue("Price")
+	fmt.Fprintf(w, "ID = %d\n", ID)
```
Comment (jboursiquot) [inline main.go]: Caution: You're using `%d` for digits but are passing in `ID` which is a string.

File: main.go [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+	fmt.Fprintf(w, "Name = %s\n", Name)
+	fmt.Fprintf(w, "Description = %s\n", Description)
+	fmt.Fprintf(w, "Price = %d\n", Price)
```
Comment (jboursiquot) [inline main.go]: Caution: You're using %d for digits but are passing in ID which is a string.

File: main.go [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+
+}
+
+func DeleteHandler(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	id := vars["id"]
+	intid := 0
+	_, err := fmt.Sscan(id, &intid)
+	if err != nil {
+		log.Fatalln(err)
```
Comment (jboursiquot) [inline main.go]: Prefer a simple `strconv.Atoi` here over the use of `fmt.Sscan`.
Do not `Fatalln` out of a handler. You'll be crashing your entire server over an invalid request parameter. Use the response writer to send back an error/status to the client.

File: main.go [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+	}
+	for _, products := range product {
+		if intid > len(product) {
+			w.WriteHeader(http.StatusNotFound)
+			response := "404 status not found"
+			_, err = w.Write([]byte(response))
```
Comment (jboursiquot) [inline main.go]: This `err` is never checked.

File: main.go [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+	}
+	response := "This is a put request"
+	_, err = w.Write([]byte(response))
+	fmt.Println("r now", r)
+	fmt.Println("r.Form", r.Form)
+	fmt.Println("r.PostForm", r.PostForm)
+
+}
+
+func DeleteHandler(w http.ResponseWriter, r *http.Request) {
```
Comment (jboursiquot) [inline main.go]: You have a logic error in your `DeleteHandler`. I've submitted the same ID for deletion multiple times and the product is never actually removed.

File: main.go [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+		log.Fatalln(err)
+	}
+	for _, products := range product {
+		if intid > len(product) {
+			w.WriteHeader(http.StatusNotFound)
+			response := "404 status not found"
+			_, err = w.Write([]byte(response))
+			return
+		}
+		if products.ID == intid {
```
Comment (jboursiquot) [inline main.go]: You need to actually remove the product from the `products` slice, not blank it out.

File: main.go [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+	Name := r.FormValue("Name")
+	Description := r.FormValue("Description")
+	Price := r.FormValue("Price")
+	fmt.Fprintf(w, "ID = %d\n", ID)
+	fmt.Fprintf(w, "Name = %s\n", Name)
+	fmt.Fprintf(w, "Description = %s\n", Description)
+	fmt.Fprintf(w, "Price = %d\n", Price)
+
+}
+func PutHandler(w http.ResponseWriter, r *http.Request) {
```
Comment (jboursiquot) [inline main.go]: This does not update any product. I assume you'll add this next.

File: main.go [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+				log.Print("error 3")
+				w.WriteHeader(http.StatusInternalServerError)
+			}
+			w.WriteHeader(http.StatusOK)
+
+			_, err = w.Write(response)
+		}
+	}
+}
+func PostHandler(w http.ResponseWriter, r *http.Request) {
```
Comment (jboursiquot) [inline main.go]: This does not add a product yet. I assume you'll add this next.

File: main.go [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+		}
+		if products.ID == intid {
+			response, err := json.Marshal(product[intid-1])
+			if err != nil {
+				log.Print("error 3")
+				w.WriteHeader(http.StatusInternalServerError)
+			}
+			w.WriteHeader(http.StatusOK)
+
+			_, err = w.Write(response)
```
Comment (jboursiquot) [inline main.go]: This `err` is never checked.

File: main.go [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+	intid := 0
+	_, err := fmt.Sscan(id, &intid)
+	if err != nil {
+		log.Fatalln(err)
+	}
+	for _, products := range product {
+		if intid > len(product) {
+			w.WriteHeader(http.StatusNotFound)
+			response := "404 status not found"
+			_, err = w.Write([]byte(response))
```
Comment (jboursiquot) [inline main.go]: This `err` is never checked.

**Review by jboursiquot — CHANGES_REQUESTED** [review]
Good start. Still needs work. In addition to inline comments, fix the following:

1. You're `.gitignore` file needs to be at the root of your repo, not inside of `.idea`
2. Recall Monday's class discussion and the reason why you do not commit your `.idea` folder.

**Comment by thottel64** [comment]
Made a mistake. Code is not ready for review.

**Comment by thottel64** [comment]
Code should now be ready for review

File: go.mod [hunk 12]
```diff
@@ -0,0 +1,5 @@
+module github.com/thottel64/pivotassignments/productserver
```
Comment (jboursiquot) [inline go.mod]: Your module is `github.com/thottel64/pivotassignments`. Remove the `productserver` portion.

File: productserver/main.go [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+	}
+}
+
+func GetIDHandler(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	id := vars["id"]
+	intid := 0
+	_, err := fmt.Sscan(id, &intid)
+	if err != nil {
+		log.Fatalln(err)
```
Comment (jboursiquot) [inline productserver/main.go]: Do not crash your server with `log.Fatalln`. Return an error back to the client instead.

File: productserver/main.go [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+	}
+	for _, products := range product {
+		if intid > len(product) {
+			w.WriteHeader(http.StatusNotFound)
+			response := "404 status not found"
+			_, err = w.Write([]byte(response))
+			if err != nil {
+				w.WriteHeader(http.StatusInternalServerError)
+			}
```
Comment (jboursiquot) [inline productserver/main.go]: Not necessary. You're already sending the appropriate response in the header.

File: productserver/main.go [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+func GetIDHandler(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	id := vars["id"]
+	intid := 0
+	_, err := fmt.Sscan(id, &intid)
+	if err != nil {
+		log.Fatalln(err)
+	}
+	for _, products := range product {
+		if intid > len(product) {
```
Comment (jboursiquot) [inline productserver/main.go]: You can do this check outside of the loop. No need to ask the same question multiple times with each loop iteration.

File: productserver/main.go [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+		}
+		if products.ID == intid {
+			response, err := json.Marshal(product[intid-1])
+			if err != nil {
+				log.Print("error 3")
+				w.WriteHeader(http.StatusInternalServerError)
+			}
+			w.WriteHeader(http.StatusOK)
+
+			_, err = w.Write(response)
```
Comment (jboursiquot) [inline productserver/main.go:88]: You capture `err` but you don't check it.
You can omit capturing the error here.

File: productserver/main.go [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+	if err := r.ParseForm(); err != nil {
+		fmt.Fprintf(w, "ParseForm() err: %v", err)
+		return
+	}
+	var newProduct Product
+	reqBody, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+	}
+
```
Comment (jboursiquot) [inline productserver/main.go]: defer the body's close

File: productserver/main.go [hunk 18]
```diff
@@ -0,0 +1,175 @@
...
+	err = json.Unmarshal(reqBody, &newProduct)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+	}
+	product = append(product, newProduct)
+	w.WriteHeader(http.StatusCreated)
+
+	json.NewEncoder(w).Encode(newProduct)
```
Comment (jboursiquot) [inline productserver/main.go]: No need to send the new product back after creating it.

File: productserver/main.go [hunk 19]
```diff
@@ -0,0 +1,175 @@
...
+func PutHandler(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	id := vars["id"]
+	intid := 0
+	_, err := fmt.Sscan(id, &intid)
+	if err != nil {
+		log.Fatalln(err)
+	}
+	for index, products := range product {
+		if intid > len(product) {
```
Comment (jboursiquot) [inline productserver/main.go]: No need to repeat this check with every loop iteration.

File: productserver/main.go [hunk 20]
```diff
@@ -0,0 +1,175 @@
...
+)
+
+type Product struct {
+	ID          int    `json:"id"`
+	Name        string `json:"name"`
+	Description string `json:"description"`
+	Price       int    `json:"price"`
+}
+
+var product []Product
```
Comment (jboursiquot) [inline productserver/main.go]: `product` is a slice of products. Use a plural variable name.

File: productserver/main.go [hunk 21]
```diff
@@ -0,0 +1,175 @@
...
+func DeleteHandler(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	id := vars["id"]
+	intid := 0
+	_, err := fmt.Sscan(id, &intid)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+	}
+	for index, products := range product {
+		if intid > len(product) {
```
Comment (jboursiquot) [inline productserver/main.go]: No need for this check in every loop iteration.

File: productserver/main.go [hunk 22]
```diff
@@ -0,0 +1,175 @@
...
+			_, err = w.Write([]byte(response))
+			if err != nil {
+				w.WriteHeader(http.StatusInternalServerError)
+			}
+			return
+		}
+		if products.ID == intid {
+			product = append(product[:index], product[index+1:]...)
+			response := "Product Deleted"
+			_, err = w.Write([]byte(response))
```
Comment (jboursiquot) [inline productserver/main.go]: No need for this response. Communicate with the right response header status code.

**Review by jboursiquot — CHANGES_REQUESTED** [review]

File: productserver/main.go (hunk 20 above)
Comment (jboursiquot) [reply productserver/main.go]: This is still unaddressed.

**Review by jboursiquot — CHANGES_REQUESTED** [review]
You have an unaddressed change request.

File: productserver/main.go (hunk 20 above)
Comment (thottel64) [reply productserver/main.go]: Sorry about that! I've renamed the variable to make the code easier to read.

**Review by jboursiquot — APPROVED** [review]

**Pull request merged**